### ISSUE TITLE
Bug: sam build omits Dockerfile suffix from build.toml when aws:asset:property == "Code.ImageUri"

### ISSUE THREAD

**Issue by christiangaetano** [aws/aws-sam-cli#4355]
### Description:

Hi all! I was unable to find an open or closed issue describing this exact problem.

The problem is very specific: when your SAM template specifies a Dockerfile containing a suffix in its name (e.g., `.production`) AND the `aws:asset:property` is set to `"Code.ImageUri"`, the Dockerfile suffix is not copied over to the generated `build.toml`, and thus SAM is unable to locate the Dockerfile. This occurs for both the `aws:asset:dockerfile-path` and `Dockerfile` properties in the template JSON.

This is important because this is the pathway by which CDK generates SAM templates which can be used for local testing and other deployment internals.

The problem boils down to the metadata normalization when `aws:asset:property` is `Code.ImageUri`. This leads to `_extract_image_asset_metadata` being called in which only the Dockerfile path's `stem` is extracted ([code permalink](https://github.com/aws/aws-sam-cli/blob/d634991615476a3980496694f90ac6c23800ab73/samcli/lib/samlib/resource_metadata_normalizer.py#L189)).

It seems to me that `dockerfile_path.name` should be used here instead of `dockerfile_path.stem` ([Python docs](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.name)). Is there some other rationale for using `.stem` here?

### Steps to reproduce:

1. Create a CDK project (e.g., [CDK hello world app](https://docs.aws.amazon.com/cdk/v2/guide/hello_world.html))
2. Add a simple Dockerfile **named `Dockerfile.production`**
3. Define a CDK Docker Lambda function. E.g.:

```typescript
new lambda.DockerImageFunction(
      this,
      "TestFunction",
      {
        code: lambda.DockerImageCode.fromImageAsset(
          path.join(__dirname, "..", "path", "to", "dockerfile"),
          {
            file: "Dockerfile.production",
          },
        ),
      },
    );
```

5. Change the name of your `Dockerfile` to `Dockerfile.prod`. Also update this reference in the template JSON/YAML.
6. `npx cdk synth --no-staging`
7. `sam build -t cdk.out/HelloCdk.template.json`
8. `sam local invoke`

### Observed result:

The build will "succeed" but you should not see a Docker build triggered. Open `.aws-sam/build.toml` and examine the function build definition metadata. You should see

```
"aws:asset:dockerfile-path" = "Dockerfile.production"
```

and

```
Dockerfile = "Dockerfile"
```

Upon attempting to invoke, you will see that SAM is unable to locate the Docker image to invoke the function (because it has not been built).

### Expected result:

The outputted `build.toml` should include:

```
Dockerfile = "Dockerfile.production"
```

`sam build` and `sam local invoke` should trigger a build of the Docker image and invoke the function by running that image, respectively.

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. macOS 13.0
3. `sam --version`: SAM CLI, version 1.61.0
9. AWS region: local development

**Comment by sriram-mv** [comment]
Thanks for creating this issue, it does look to be a bug. Would you be willing to raise a PR for it and we will prioritize on getting it through! cc: @moelasmar

**Comment by christiangaetano** [comment]
@sriram-mv thank you for the response! Yep absolutely, I can get a PR up sometime today.

**Comment by mndeveci** [comment]
Changes have been released with [SAM CLI v1.62.0](https://github.com/aws/aws-sam-cli/releases/tag/v1.62.0)
